src/libDB/Archival.cpp

```diff
@@ -0,0 +1,221 @@
+            }
+            if (m_mediator.m_txBlockChain.GetBlockCount() != 1)
+            {
+                txBlockNum = m_mediator.m_txBlockChain.GetBlockCount();
+            }
+            LOG_GENERAL(INFO,
+                        "TxBlockNum " << txBlockNum
+                                      << " DSBlockNum: " << dsBlockNum);
+            m_synchronizer.FetchLatestDSBlocks(m_mediator.m_lookup, dsBlockNum);
+            m_synchronizer.FetchLatestTxBlocks(m_mediator.m_lookup, txBlockNum);
+            m_synchronizer.FetchLatestState(m_mediator.m_lookup);
+
+            if (m_mediator.m_currentEpochNum % NUM_FINAL_BLOCK_PER_POW == 0)
+            {
+                if (!m_mediator.m_lookup->CheckStateRoot())
+                {
+                    LOG_GENERAL(WARNING, "Archival State Root mis-match");
+                }
+            }
+            m_mediator.m_lookup->GetShardFromLookup();
+            if (m_mediator.m_currentEpochNum > 1)
+            {
+                SendFetchMicroBlockInfo();
+                SendFetchTxn();
+            }
+            if (m_mediator.m_currentEpochNum % NUM_FINAL_BLOCK_PER_POW == 0)
+            {
+                LOG_GENERAL(INFO,
+                            "Sleep for " << REFRESH_DELAY
+                                    + POW_BACKUP_WINDOW_IN_SECONDS);
+                this_thread::sleep_for(chrono::seconds(
+                    REFRESH_DELAY + POW_BACKUP_WINDOW_IN_SECONDS));
+            }
+            else
+            {
+                this_thread::sleep_for(chrono::seconds(REFRESH_DELAY));
+            }
+        }
+    };
+    DetachedFunction(1, func);
+}
+
+bool
+    Archival::Execute([[gnu::unused]] const std::vector<unsigned char>& message,
+                      [[gnu::unused]] unsigned int offset,
+                      [[gnu::unused]] const Peer& from)
+{
+    LOG_MARKER();
+    return true;
+}
+
+void Archival::Init()
+{
+    m_mediator.m_dsBlockChain.Reset();
+    m_mediator.m_txBlockChain.Reset();
+    {
+        std::lock_guard<mutex> lock(m_mediator.m_mutexDSCommittee);
+        m_mediator.m_DSCommittee->clear();
+    }
+    AccountStore::GetInstance().Init();
+
+    m_synchronizer.InitializeGenesisBlocks(m_mediator.m_dsBlockChain,
+                                           m_mediator.m_txBlockChain);
+}
+
+bool Archival::AddToFetchMicroBlockInfo(const uint64_t& blockNum,
+                                        const uint32_t shardId)
+{
+    LOG_MARKER();
+
+    lock_guard<mutex> g(m_mutexMicroBlockInfo);
+    LOG_GENERAL(
+        INFO, "Added " << blockNum << " " << shardId << " to fetch microBlock");
+    m_fetchMicroBlockInfo[blockNum].push_back(shardId);
+
+    return true;
+}
+
+bool Archival::RemoveFromFetchMicroBlockInfo(const uint64_t& blockNum,
+                                             const uint32_t shardId)
+{
+    LOG_MARKER();
+
+    lock_guard<mutex> g(m_mutexMicroBlockInfo);
+
+    vector<uint32_t>& shard_ids = m_fetchMicroBlockInfo[blockNum];
```

Comment:
Should use `contains()` to check whether the key exist otherwise it will create one with the using of `[]`, and here better to use `at()` to provide a boundary check and exception handling